### PR TITLE
Do not remove generated files from existing rules

### PR DESCRIPTION
**Pull request by gg-sr**
Before, generated sources were removed from `srcs` and `hdrs`; rules with only generated sources would be removed as well.

With this PR, `srcs` and `hdrs` keep their generated sources, and rules with only generated sources are kept.

This also seems enough for Gazelle to add dependencies to targets when generated headers are included (e.g. in the example added in this PR, a dependency from `test` to `alone` is correctly added based on `#include "generated/alone.h"`).

I decided to only preserve existing `srcs` and `hdrs`, rather than adding new ones (e.g. by adding generated sources to `fileInfos`). I figured treating generated files as sources files would be nice, but would likely have more edge cases.

I cooked this up pretty quickly. I guess there are better ways to add tests than modifying `example/bzlmod`, but this was the fastest way to test things (and worked with the debugger).

### CODE REVIEW

File: language/cc/generate.go [hunk 1]
```diff
@@ -683,6 +698,27 @@ func (info *rulesInfo) matchExistingRule(kind, name string, srcGroups sourceGrou
...
+		if info.genFiles.Contains(f) {
+			genHdrs = append(genHdrs, f)
+		}
+	}
+	for _, f := range existingRule.AttrStrings("srcs") {
+		if info.genFiles.Contains(f) {
+			genSrcs = append(genSrcs, f)
+		}
+	}
+	return
```
Comment (jayconrod) [inline language/cc/generate.go]: Please use explicit returns rather than naked returns. https://go.dev/wiki/CodeReviewComments#named-result-parameters

File: language/cc/generate.go [hunk 2]
```diff
@@ -571,6 +577,12 @@ func (c *ccLanguage) findEmptyRules(args language.GenerateArgs, fileInfos []file
 		if existingFileIndex >= 0 {
 			continue
 		}
+		if collections.ContainsFuncSeq(rulesInfo.genFiles.All(), func(name string) bool {
```
Comment (jayconrod) [inline language/cc/generate.go]: There must be a simpler way to express this. Is this equivalent to `rulesInfo.genFiles.Intersects(ruleFiles)`?

File: language/cc/testdata/generated_files/BUILD.in [hunk 3]
```diff
@@ -0,0 +1,34 @@
+load("@rules_cc//cc:defs.bzl", "cc_library", "cc_test")
```
Comment (jayconrod) [inline language/cc/testdata/generated_files/BUILD.in:1]: Please don't change the example, this is meant to be pretty minimal. Instead, add a directory test to `language/cc/testdata`.

**Review by jayconrod — COMMENTED** [review]
Thanks for sending this fix.

A few comments. Please also add `Signed-off-by: Your Name <your@email.com>` to all commit messages (use `git commit -s`).

**Comment by gg-sr** [comment]
Thank you for the review!

I had to force push to change the commit message with `--signoff`. I made commits for subsequent changes with the expectation that they'd all be squashed before merging into `main`.

Notably I added helpers to debug tests, similar to how we can debug examples. I can revert them, but I figure they could be useful to others.

Btw, I tried to add support for `:generated_file.h` syntax as well, but it seems non-trivial (removing the `:` prefix is part of the solution, but not enough) so I just dropped it.

File: language/cc/generate.go (hunk 2 above)
Comment (gg-sr) [reply language/cc/generate.go]: Indeed, didn't notice that they were both sets! Thank you.

File: language/cc/testdata/.gitignore
Comment (gg-sr) [inline language/cc/testdata/.gitignore:1]: Ah, I just realized that there _is_ a `MODULE.bazel.lock` file in this subtree. I guess I should revert this?

Comment (jayconrod) [reply language/cc/testdata/.gitignore:1]: It's fine to ignore it. The `MODULE.bazel` file is only there for Gazelle to be able to find the repo root. The lock files don't matter here.

File: .vscode/launch.json [hunk 4]
```diff
@@ -29,6 +29,26 @@
         },
       ]
     },
+    {
```
Comment (jayconrod) [inline .vscode/launch.json]: Let's leave out the test helpers change. I'd rather keep the PR focused on one purpose.

File: language/cc/testdata/.gitignore
Comment (gg-sr) [reply language/cc/testdata/.gitignore:1]: I'm not sure I understand, should I ignore the `MODULE.bazel.lock` files via `.gitignore`, or myself (and either commit it or make sure not to commit, but without `.gitignore`)?

Since they're generated automatically when running tests I think it'd be nice to either ignore them via `.gitignore` or commit them.

Comment (jayconrod) [reply language/cc/testdata/.gitignore:1]: `.gitignore` please

Comment (gg-sr) [reply language/cc/testdata/.gitignore:1]: Okay, I left it. The PR should be good to review.

**Review by jayconrod — APPROVED** [review]

**Comment by jayconrod** [comment]
As mentioned in the other PR, CI is failing because of a misconfiguration on my side: we enabled remote caching, but the secret needed to make that work isn't available on PRs from external contributors.

I've verified tests pass locally, so that's good enough for now.
